### PR TITLE
Rework Spawn System

### PR DESCRIPTION
**Pull request by adQuid**
**Brief Description of What This PR Does**

Reworks the spawn system to only spawn once per area around the player, excluding microbe spawn which is continuous. Each "sector" is about halfway to the edge of the screen when zoomed out, and new entities appear 2 or more sectors away from the player. 

Because there are 12 sectors possibly spawning microbes now, the Microbe Spawner has been modified to spawn them less often. There is also an AI change to make sure newly spawned AI is headed towards the player instead of just swimming around offscreen waiting to despawn. Because of both of these, I also took out the fancier colony spawn patterns to save on some computation time.

I also increased the cloud spawn radius to prevent clouds despawning too soon, and tied the chemoreceptor max range to that value, so that detection lines wont mysteriously go away because the cloud despawned (and to prevent excess computation)

Finally, I added just a little phosphates that always spawns at the start of the game, so that the early game isn't completely empty and to give a new player something obvious and visible to interact with.

There is still an issue with the tiny purple circle that appears then fades away sometimes, but that issue is already in master branch, and I don't know where to start fixing that.

**Related Issues**

I don't see any from a cursory scan of the issues list, but let me know if I'm wrong.

Note: This probably renders https://github.com/Revolutionary-Games/Thrive/pull/2913 obsolete

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [X] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

File: src/microbe_stage/SpawnSystem.cs
Comment (adQuid) [inline src/microbe_stage/SpawnSystem.cs:284]:
```suggestion
    ///   Does a single spawn with a spawner
```

**Comment by buckly90** [comment]
Incredible change! It makes the world feel far from barren and actually alive.

Unfortunately, there is a glaring performance issue after a few generations. Cells would just pile up and pile up until I got to about 20 fps, and sometimes even lower. And that's pretty scary with my setup.

Strangely enough, even dying, which usually despawns everything, did not fix the performance issue. Does everything despawn as intended? Is there some kind of memory leak?

File: src/microbe_stage/MicrobeAI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeAI.cs:227]: Do we want to lump this in with the spawn system change which is already a huge difference?

File: simulation_parameters/Constants.cs
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:22]: XML needed

**Comment by hhyyrylainen** [comment]
Unless the code is hidden pretty well, this seems to now ignore the maximum spawn limit of stuff, which would mean that I'd also expect the performance to tank a lot because the entity limit is really the only thing keeping the game from falling to a couple of FPS...

File: src/microbe_stage/MicrobeAI.cs
Comment (adQuid) [reply src/microbe_stage/MicrobeAI.cs:227]: I'm pretty confident this is now essential. With microbes now spawning in a ring far from the player, without this change you would need to spawn almost 5 times as many to get the same number to appear on screen.

**Comment by Oliveriver** [comment]
Had a go with this since I was curious. It does indeed make the world much more dynamic and much more consistent*.

However, that had the knock-on effect of making it much easier. I had no issue at all collecting enough compounds almost immediately every generation. The spawn system is consistent, but it's consistently high. If there is a simple way to turn down this dial, that would be great, especially since that would provide one option for difficulty variation.

I also ran into performance issues pretty quickly. Hopefully reimplementing the entity limit fixes that.

* The exception was when I entered a new patch for the first time. It was often completely barren for long enough that I ran out of compounds and died. After that though, all was well.

File: src/microbe_stage/MicrobeAI.cs
Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeAI.cs:227]: Wasn't the idea with sector spawns that sectors contain specific amount of stuff, which would even load in and out as the player moves? Did this PR fundamentally change from the first spawn system PR?

Comment (adQuid) [reply src/microbe_stage/MicrobeAI.cs:227]: I'm not intimately familiar with the other PR, but based on what you're describing this is a much simpler solution. All this change does is take the "Biased circle" spawn pattern and instead track what areas have and have not spawned something in them. It then relies on existing despawn functionality to remove things too far from the player.

This works fine for chunks and clouds which don't move, but most microbes won't just sit around and wait to be discovered; they will start moving, usually out of the perimeter for depsawning and never enter the game properly. My suspicion is that this is already an issue in master, but now that the Microbes are spawning a little further out (so they don't appear onscreen) it's a more noticeable problem.

**Comment by adQuid** [comment]
> If there is a simple way to turn down this dial, that would be great, especially since that would provide one option for difficulty variation.

This variable (which I just moved into the constants file) would be the easiest way to do that, and not a bad way to practice messing with the code :) https://github.com/Revolutionary-Games/Thrive/pull/3235/files#diff-1866e717ab370a48530eaa34ab83ade4adf2d08d2a63befb6adb2d8c5e951211R27

File: src/microbe_stage/MicrobeAI.cs
Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeAI.cs:227]: Ah okay, you should rename this PR to not contain the "sector" word in the name then.

**Comment by hhyyrylainen** [comment]
@adQuid you should still use the `Constants.DEFAULT_MAX_SPAWNED_ENTITIES` constant for the entity limit, or rename / remove that previous one if it isn't used anymore.

**Comment by hhyyrylainen** [comment]
I think a pretty huge finding has been made regarding the spawn system, so I think this PR should wait for that to be complete and then we should re-evaluate this: https://github.com/Revolutionary-Games/Thrive/pull/3243

**Comment by adQuid** [comment]
Understood. That works out great for me since I won't have much time to work on this until the weekend anyway.

**Comment by adQuid** [comment]
There is still an unfortunate tendency for islands of compound clouds to form, I think due to limitations in the current compound cloud system, but the lag spikes seem to be fixed and I like the microbe spawn rates a lot better now.

**Comment by Kemikal1** [comment]
Did some testing and to me the compund islands werent so obvious. It does spawn more microbes indeed maybe around 70-100 entitites from 20-30 normally so more lively definetly.

**Comment by buckly90** [comment]
Performance is much better now. My fps never dropped below the 30-40 range which still might not be good given I have a pretty good PC, but it's at least playable now.

For some reason it feels like compounds spawns are far less consistent since I last tested. There would be periods of time where I would just swim and swim and encounter maybe one or two clouds. This issue seemingly became more frequent as time went on. I even had to double check to make sure I was on the right branch because it felt a bit too similar to what we already have. Microbe spawns are at least rather frequent so I was able to survive off of predation.

**Comment by RacerBG** [comment]
> Finally, I added just a little phosphates that always spawns at the start of the game, so that the early game isn't completely empty and to give a new player something obvious and visible to interact with.

Is it a good idea though? If the tutorial is enabled, it should guide the player towards the nearest cloud of glucose.

Otherwise, I agree with the changes you made but broader testing is needed.

**Comment by adQuid** [comment]
> Is it a good idea though? If the tutorial is enabled, it should guide the player towards the nearest cloud of glucose.

I'm really confident this will be important. A side effect of the new spawn system not spawning stuff in view of the player is that the screen starts out blank save for one microbe. If the player follows the tutorial and just swims in circles or chooses to skip it (I assume most do, people like to skip those) there are very few early clues about what the gameplay is about, and a player could reasonably wonder if the cell is moving at all or the background is just drifting. Look up any talk on video game tutorials, and you'll hear the same thing: don't throw a wall of text at the player, give them something to experiment with. The start of the game is a great time to have one cell, one cloud directly to the right, and a quick demonstration of what the first 10 minutes of the game are like.

And if I'm wrong, it's one purple puff that fills up half the phosphate bar.

**Comment by adQuid** [comment]
@buckly90 Having confirmed that the issue I was talking about on Discord was a fluke, I now think that this is in about the best state I can get it into without a massive compound cloud refactor, which I just don't have the time to commit to right now.

File: simulation_parameters/Constants.cs
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:391]: If we assume that cloud spawn isn't equal to the compound cloud plane distance, restricting this distance doesn't seem like the best idea to me...

File: src/microbe_stage/MicrobeAI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeAI.cs:77]: Shouldn't this be saved? If an old save doesn't have it, this will be false. But for newer saves this seems like a good thing to save.

File: src/microbe_stage/MicrobeStage.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:860]: This should not be required with the despawn fixes... if this is required please add a comment detailing why this is needed. I think this'll make things visually pop out of existance on camera so I think things will look better if we don't do this and instead rely on the more powerful despawn system to be able to get things done.

**Comment by Leinourdian** [comment]
I did some testing out of curiosity. I like the feel of the new system a lot. It does indeed feel livelier. I also like that things spawn in already before you swim into a new area so I don't have to swim in big circles to get more spawns. My only problem at the moment is the performance which is worse than in 5.8.1. I think it's caused by two things: Firstly, there are more entities in general. Secondly, entities can spawn in quick bursts.

When it comes to the first one, I could sometimes have up to 140 entities (the entity limit), but I'd only see 5 to 30 entities on the screen with max zoom out. That means that about 80-95% of entities were off-screen where the player doesn't really care about them. I'm wondering if it'd be possible to do something to reduce the amount entities off-screen? Or perhaps off-screen entities could be made less performance heavy because they probably don't have to be fully simulated. That's probably easier said than done but it could help performance a lot.

As for the second one, I think that the spawn system can at worst spawn up to dozens of entities in a single frame. You can sometimes see the entity count jump up really fast. I saw some smallish lag spikes that were probably caused by it. I think this is because `spawnsLeftThisFrame` is not taken into consideration inside `SpawnEntities` or the methods used by it like `SpawnWithSpawner` unlike how it used to be.

I noticed the despawning of all entities on player death by the way. I couldn't usually see it but a low framerate sometimes made it visible for a fraction of a second. It didn't look that weird but it's something to take into consideration.

`Denstity` should probably be `Density`.
https://github.com/Revolutionary-Games/Thrive/blob/127c12282e8881290f7ee82e90209f56da88ce7d/src/microbe_stage/Spawner.cs#L30

**Comment by hhyyrylainen** [comment]
> As for the second one, I think that the spawn system can at worst spawn up to dozens of entities in a single frame. You can sometimes see the entity count jump up really fast. I saw some smallish lag spikes that were probably caused by it. I think this is because spawnsLeftThisFrame is not taken into consideration inside SpawnEntities or the methods used by it like SpawnWithSpawner unlike how it used to be.

Sounds like a new bug with this PR. The current spawn system definitely should respect the spawns per frame.

**Comment by Leinourdian** [comment]
> Sounds like a new bug with this PR. The current spawn system definitely should respect the spawns per frame.

Yeah, it's new to this PR.

File: simulation_parameters/Constants.cs
Comment (adQuid) [reply simulation_parameters/Constants.cs:391]: This really was an attempt to fix a related but unnecessary issue anyway.

**Comment by Oliveriver** [comment]
I think my feedback remains about the same as it was: a much more consistent and livelier gameplay experience, but still too easy and still hitting some major performance issues.

I tried passing the same save between master and this branch a few times. In master, I never saw more than 40 entities, and almost never dropped frames. In this branch, I consistently had 80+ entities, sometimes hitting 130, and had both lag spikes and prolonged periods of 30-40 FPS.

**Comment by Kemikal1** [comment]
i think the fact that 130 entities create such a lag is a problem in itself.

**Comment by hhyyrylainen** [comment]
If someone can fix the performance problems with many entities, I would be extremely pleased. In the meantime we must limit the number of entities we allow in the game.

File: src/microbe_stage/MicrobeStage.cs
Comment (adQuid) [reply src/microbe_stage/MicrobeStage.cs]: I've moved this in order to keep the entities from disappearing while the player might see, but the despawn definitely is needed with this spawning system, since without it the game will hold twice as many entities as the max for a couple seconds.

**Comment by buckly90** [comment]
No noticeable drops in framerate, and the world looks livelier than ever!
I never experienced any sudden fields of nothing, there was always a resource nearby, useful or not.

Unfortunately, it seems the density of ammonia and phosphate is a bit too low. It spawns often and consistently enough, but in amounts that contribute little to progress. The result is a massively stretched amount of time between generations. This wouldn't actually be a problem if there was more engagement in the gameplay phase, but for now my suggestion is to tune up the density. You can leave that up to me if you like.

You've done a wonderful job here, this PR functions beautifully from a gameplay standpoint.

**Comment by Oliveriver** [comment]
Misclicked while looking at this on mobile and accidentally merged master into this PR. Sorry about that...

**Comment by adQuid** [comment]
meh, I was probably about to do that anyway

File: simulation_parameters/Constants.cs
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:23]:
```suggestion
    ///   Size of "chunks" used for spawning entities
```

File: src/microbe_stage/PatchManager.cs
Comment (hhyyrylainen) [inline src/microbe_stage/PatchManager.cs:162]: `OrderByDescending`?

File: src/microbe_stage/Spawners.cs
Comment (hhyyrylainen) [inline src/microbe_stage/Spawners.cs:105]: With the playerPosition no longer being in here, does this method still make sure it doesn't spawn the colony members so far away that they can despawn instantly?

File: src/microbe_stage/PatchManager.cs
Comment (hhyyrylainen) [inline src/microbe_stage/PatchManager.cs:219]: What happened to this spawn frequency updating code?

File: src/microbe_stage/MicrobeStage.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:390]: Is there a chance that this crashes the game? There was that one random guy who tried to crank up the despawn rate a ton and seemed to get more crashes that way. Anyone experienced crashes in testing?

File: src/microbe_stage/Spawners.cs
Comment (adQuid) [reply src/microbe_stage/Spawners.cs:105]: No, that PR was a victim of bad timing. Colonies now have a max size of 2, and spawn well within the despawn radius.

File: src/microbe_stage/PatchManager.cs
Comment (adQuid) [reply src/microbe_stage/PatchManager.cs:219]: I admit I don't recall exactly since It's been a while since I made this change, but I do know that frequency and density were combined into one value, since they were just scalar multiples of each other anyway.

File: src/microbe_stage/MicrobeStage.cs
Comment (adQuid) [reply src/microbe_stage/MicrobeStage.cs:390]: @buckly90 @Nunez2196 @Oliveriver ?

Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeStage.cs:390]: It *might* help a bit to use the `Invoke.Instance.Queue` to trigger this on the next frame as on this frame the player is already spawned, so that would spread things out a bit over different frames, though I'm not sure if that has any effect on anything...

Comment (Oliveriver) [reply src/microbe_stage/MicrobeStage.cs:390]: Nothing close to a crash for me.

Comment (adQuid) [reply src/microbe_stage/MicrobeStage.cs:390]: I could certainly add more stuff here if you want me to, but obviously with no crashes so far I can't prove that it's doing anything.

Comment (Nunez2196) [reply src/microbe_stage/MicrobeStage.cs:390]: i didn't experience any crashes either

File: src/microbe_stage/PatchManager.cs
Comment (hhyyrylainen) [reply src/microbe_stage/PatchManager.cs:219]: Are they still correctly updated, though? This even has the logging for the spawn frequency change, so my worry is that your changes actually break the spawners from actually updating...

Comment (adQuid) [reply src/microbe_stage/PatchManager.cs:219]: It is not, which is probably what accounted for the overrepresentation Nunz noticed. I've pushed a fix to correct.

Comment (hhyyrylainen) [reply src/microbe_stage/PatchManager.cs:219]: Okay so we found a bug here. @Nunez2196 can you re-check that now?

**Comment by Oliveriver** [comment]
I don't want to be the bearer of bad news, but I just tested this again and ended up in a patch with zero compound clouds. Plenty of other cells, but no clouds at all for at least six death and respawn cycles. I can provide logs if needed.

**Comment by adQuid** [comment]
@Oliveriver Found it. I failed to pass on a strange multiplication we do (and I contributed to, the cloud system as a whole is due for a refactor I won't attempt), resulting in cloud density decreasing by 10,000 every generation.

File: src/microbe_stage/SpawnSystem.cs
Comment (hhyyrylainen) [inline src/microbe_stage/SpawnSystem.cs:347]: There seems to be a massive bug here, this no longer limits the spawns per frame properly here, at least I can't see this loop exiting if there are no spawns left. Was the lag spikes the reason why you removed the colony spawns or? If so then, here's the bug which started to cause the lag spikes.

Also now that you've broken the queued spawns system, the left in code to spawn queued things, is completely useless... so I very strongly say you should fix the queued spawns system to ensure no minor stuttering happens from spawning more than a couple of things per frame.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
I've now reviewed all the files and found some things to comment on in the SpawnSystem.cs file, including probably a bug that this broke the max spawns per frame system.

File: src/microbe_stage/SpawnSystem.cs
Comment (adQuid) [reply src/microbe_stage/SpawnSystem.cs:270]: I don't know an intuitive pair of nested loops that returns this very unusual pattern of 4 lines of squares, but preseting the list size makes sense.

Comment (hhyyrylainen) [reply src/microbe_stage/SpawnSystem.cs:270]: Maybe then this could be made into 4 loops? That'd still reduce the repetition a tiny bit...

File: src/microbe_stage/SpawnSystem.cs
Comment (adQuid) [reply src/microbe_stage/SpawnSystem.cs:55]: I think the fate of this variable will depend on our conclusion about queues.

File: src/microbe_stage/SpawnSystem.cs
Comment (adQuid) [reply src/microbe_stage/SpawnSystem.cs]: Just didn't know about Int2. Fixed

File: src/microbe_stage/SpawnSystem.cs
Comment (hhyyrylainen) [reply src/microbe_stage/SpawnSystem.cs:358]: I fail to understand why you think that using the spawn queue would impact the sectors. The logic should be:
- If there are queued spawns, handle those
- If out of spawns this frame, skip
- Otherwise enter into the spawn sectors code. When spawning there, spawn up to spawnsLeftThisFrame things and if there are more stuff to spawn, then save that as the queue



It's simply not doable to spawn more things than spawnsLeftThisFrame, it's a massive source of small lag spikes, which I don't want to put back into the game. If you have changed the logic so that multiple spawners are allowed to trigger per frame, you need to also change the spawn queueing system to allow multiple spawners to have active queues at once, then no spawns that are wanted will be forgotten.

Comment (adQuid) [reply src/microbe_stage/SpawnSystem.cs:358]: Unless I'm deeply misunderstanding something (and if so educate me now) queuedSpawns is not a list of QueuedSpawn, it's a single nullable instance. That means the queue has a max length of 1. That's not enough if multiple things can spawn per frame.

Comment (hhyyrylainen) [reply src/microbe_stage/SpawnSystem.cs:360]: It is a nullable *enumerator* and an enumerator can return as many items as it wants. If you do `foreach(var thing in list)` that uses an enumerator. The list is converted into an enumerator and then the foreach loops it. So basically we have a nullable reference to a thing that promises to generate us one or more items we can read one at a time.

Comment (adQuid) [reply src/microbe_stage/SpawnSystem.cs:360]: I see how that works within a single spawntype (spawning a colony), but this code: https://github.com/Revolutionary-Games/Thrive/blob/master/src/microbe_stage/SpawnSystem.cs#L391 is not appending anything, it's replacing whatever was in the variable. This would be very hard to test, but I'm pretty sure that during a lag spike that will remove the queues from any spawn types but the last one to be run.

Comment (hhyyrylainen) [reply src/microbe_stage/SpawnSystem.cs:360]: I'm 99% certain that that code cannot run if the previous spawner still has stuff. Look at like 193 in Process. And unless there's a bug in HandleQueuedSpawns, it won't report that there are any spawns left if it didn't manage to spawn the entire queue.

**Comment by Kasterisk** [comment]
Does this close https://github.com/Revolutionary-Games/Thrive/issues/2055 or is that a completely different approach?

**Comment by hhyyrylainen** [comment]
Yeah, this does not address #2055 so that's still something to consider in the future even after merging this.

**Comment by adQuid** [comment]
I've modified the spawn queue to be able to support multiple spawn types. I've also got a linter error that seems incorrect, which is next for me to look at.

**Pull request merged**
